File: CodinGame/Community/magic_stones.py
#!/usr/bin/env python3

# NB: Given input numbers happen to be actually so small that this can
#     reasonably be interpreted as a binary addition. E.g. in Ruby:
# n = gets.to_i
# puts gets.split.map{|x| 1<<x.to_i}.sum.to_s(2).count('1')

from heapq import *

N = int(input())
H = list(map(int,input().split()))
heapify(H)

res = 0
while H:
    i = heappop(H)
    if H and H[0]==i:
        heappop(H)
        heappush(H,i+1)
    else:
        res += 1
print(res)

File: CodinGame/Community/magic_stones.rb
#!/usr/bin/env ruby

n = gets.to_i
a = gets.split(" ").map(&:to_i).sort.chunk{|x| x}.map{|x,l| [x,l.size]}
res = 0
until a.empty? do
    x,l = a.shift
    if l>1
        if !a.empty? && a[0][0]==x+1
            a[0][1] += l/2
        else
            a.unshift([x+1,l/2])
        end
    end
    res += l%2
end
puts res
